tools/src/main/scala/scala/scalanative/build/NativeLib.scala

```diff
@@ -1,27 +1,95 @@
 package scala.scalanative
 package build
 
+import java.io.File
 import java.nio.file.Path
+import java.util.regex._
 
 /** Original jar or dir path and generated dir path for native code */
 private[scalanative] case class NativeLib(src: Path, dest: Path)
 
+/** Utilities for dealing with native library code */
 private[scalanative] object NativeLib {
+  private val fs           = File.separator
+  private val jarExtension = ".jar"
 
-  /** List of source patterns used */
-  val jarExtension  = ".jar"
-  val srcExtensions = Seq(".c", ".cpp", ".S", ".h", ".hpp")
-  val srcPatterns   = srcExtensions.mkString("glob:**{", ",", "}")
+  /** Object file extension: ".o" */
+  val oExt = ".o"
+
+  /** C++ file extension: ".cpp" */
+  val cppExt = ".cpp"
+
+  /** List of source patterns used: ".c, .cpp, .S" */
+  val srcExtensions = Seq(".c", cppExt, ".S")
+
+  /**
+   * Name of directory that contains native code: "scala-native"
+   */
+  val codeDir = "scala-native"
+
+  /** Used to find native source files in directories */
+  val srcPatterns = srcExtensions.mkString(
+    s"glob:**${fs}classes${fs}${codeDir}${fs}**{",
+    ",",
+    "}")
+
+  /** Used to find native source files in jar files */
+  private val jarSrcRegex: String = {
+    val regexExtensions = srcExtensions.mkString("""(\""", """|\""", ")")
+    s"""^${codeDir}${fs}(.+)${regexExtensions}$$"""
+  }
+
+  /**
+   * This method guarantees that only code copied and generated
+   * into the `native` directory and also in the `scala-native`
+   * sub directory gets picked up for compilation.
+   *
+   * @param workdir the base working directory
+   * @return the source pattern
+   */
+  def destSrcPatterns(workdir: Path): String = {
+    val workdirName = workdir.getFileName()
+    srcExtensions.mkString(
+      s"glob:**${fs}${workdirName}${fs}*${fs}${codeDir}${fs}**{",
+      ",",
+      "}")
+  }
+
+  /**
+   * Allow all the object files ".o" to be found with one
+   * directory recursion.
+   *
+   * @param workdir the base working directory
+   * @return the object file pattern
+   */
+  def destObjPatterns(workdir: Path): String = {
+    val workdirName = workdir.getFileName()
+    s"glob:**${fs}${workdirName}${fs}*${fs}${codeDir}${fs}**${oExt}"
+  }
 
   /** To positively identify nativelib */
-  val nativeLibMarkerFile = "org_scala-native_nativelib.txt"
+  private val nativeLibMarkerFile = "org_scala-native_nativelib.txt"
 
-  val dirMarkerFilePattern = "glob:**" + nativeLibMarkerFile
+  /** Note: assumes that code is compiled into the `classes` directory */
+  private val dirMarkerFilePattern =
+    s"glob:**${fs}classes${fs}" + nativeLibMarkerFile
```

Comment:
Another `classes` that should not be there.

Reply:
RIght, but that is more specific. I don't expect anyone to have a text file with that name. This is related to the discussion above about `classes`.

Reply:
I use the full path here as well to be explicit. It could be `s"glob:**${nativeLibMarkerFile}"` but then we can drop the `**`.

Reply:
Fixed as well.